include/natalie/constants.hpp

```diff
@@ -1,5 +1,8 @@
 // This is used at build time as well in order to generate numbers.rb
 #include <limits.h>
+#include <math.h>
 
 #define NAT_MAX_FIXNUM LLONG_MAX
-#define NAT_MIN_FIXNUM LLONG_MIN + 1
\ No newline at end of file
+#define NAT_MIN_FIXNUM LLONG_MIN + 1
+
+const long long NAT_MAX_FIXNUM_SQRT = (long long)sqrt((double)NAT_MAX_FIXNUM);
```

Comment:
Hm clang does not like this (me neither). There does not seem to be a way to define a compile time sqrt without having a custom sqrt function?

Reply:
GCC allows this,
Clang might allow it with `-fno-math-errno` set, but not sure
also the arguments to sqrt are always* doubles, so that cast seems redundant

*: In some libcs it can also do float->float and Integral-> double implicitly without the any casting or suffix'
[see here](https://en.cppreference.com/w/cpp/numeric/math/sqrt)

Reply:
I've just added the cast after the first test run failed, where clang complained about NAT_MAX_FIXNUM not being a double but a long long. Not sure if we want to add this flag.. Maybe we can find another way around this...

Reply:
@seven1m What do you think we should do with this? I guess we have the following options:

* Try this `-fno-math-errno` flag for clang that @Hendiadyoin1 suggested
* Remove this constant (included the optimization I added in Integer multiplication overflow checks)
* Implement a custom sqrt function that is a defined as constexpr
* Calculate the sqrt ourselves (3037000499) and put a constant value into the macro (this would mean we lose the portability of using LLONG_MAX)

Maybe there are even other ways around it? I'm not really a C++ dev so I have no clue if there are other options around.

Reply:
what if we take a slightly different approach? 

---- Before you continue reading this, I might be saying something extremely dumb, in which case just let me know :D 

What we need is not the sqrt of the max value, but rather the last number that when elevated to the pow of 2 will not overflow. This can be prechecked and runs reasonably fast, in fact you only need sizeof(nat_int_t) (or whatever our fixnum type is) iterations to calculate that.

If you start from the highest bits you can check if setting that bit to 1 will make the pow^2 of it overflow, if not then set it to 1, otherwise leave it to 0, step onto the next lower bit and repeat until you reach the first digit. Something like that would be:
- easy to implement
- fast
- easy to read 

and thus we could run it as a constexpr without making the compilation any slower (not that I would mind anyway, I mostly like the fact that it is somewhat simple to read/write in this case).

What do you think?

Reply:
Just for my understanding: The calculation you described would result in the sqrt of LLONG_MAX right? It's the same number because the sqrt is the last number that can be squared without overflowing.

I guess this would work though. How would you do the overflow check? Just using the overflow check we use normally?

Reply:
>  The calculation you described would result in the sqrt of LLONG_MAX right?
it'd get you the closest integer number there is to it yeah :) 

> How would you do the overflow check? 
I'd write it something like this:

```c++
nat_int_t mask = 1 << (sizeof(nat_int_t) - 1); 
nat_int_t max_sqrt = 0;

while (mask > 0) {
   nat_int_t current_sum = max_sqrt + mask;
   if ((NAT_MAN_FIXNUM / current_sum) >= current_sum) continue; // overflow
   max_sqrt = current_sum;
   mask >>= 1;
}

return max_sqrt;

```

wdyt @richardboehme ?

Reply:
Looks great! Could you have an eye on the small changes I needed to do to your code snippet?

Reply:
Looks good!